In [2]:
import pandas as pd
import numpy as np
import gzip
import json

## **Carga de datasets**

In [3]:
#Cargamos los datasets en df de pandas
steam_games = pd.read_json("steam_games.json.gz", lines=True)
steam_games['price'].astype('str')
steam_games

publisher                                 genres   
0                  None                                   None  \
1                  None                                   None   
2                  None                                   None   
3                  None                                   None   
4                  None                                   None   
...                 ...                                    ...   
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444             None                                   None   

                        app_name                     title   
0                           None                      None  \
1                           None                      None   
2                           None                      None   
3                           None                      None   
4                           None                      None   
...                          ...                       ...   
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                      None   

                                                      url release_date   
0                                                    None         None  \
1                                                    None         None   
2                                                    None         None   
3                                                    None         None   
4                                                    None         None   
...                                                   ...          ...   
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...         None   

                                                     tags   
0                                                    None  \
1                                                    None   
2                                                    None   
3                                                    None   
4                                                    None   
...                                                   ...   
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url   
0                                                    None  \
1                                                    None   
2                                                    None   
3                                                    None   
4                                                    None   
...                                                   ...   
120440  http://steamcommunity.com/app/773640/reviews/?...   
120441  http://steamcommunity.com/app/733530/reviews/?...   
120442  http://steamcommunity.com/app/610660/reviews/?...   
120443  http://steamcommunity.com/app/658870/reviews/?...   
120444  http://steamcommunity.com/app/681550/reviews/?...   

                                                    specs price  early_access   
0                                       

In [4]:
# Para este archivo y el siguiente abrimos el archivo con la libreria de gzip para leerlo en binario,
# luego lo iteramos y lo normalizamos como json con ast (ya que los json estan con comillas simples)
#y lo guardamos en una lista para luego transformarlo en Dataframe
import ast
import gzip
rows = []

with gzip.open("user_reviews.json.gz", mode='rt', encoding="utf-8") as archivo:
    for linea in archivo:
        try:
            obj = ast.literal_eval(linea)
            rows.append(obj)
        except json.JSONDecodeError:
            print(f'Error en la linea {linea}')

user_reviews = pd.DataFrame(rows)

user_reviews

user_id                                           user_url   
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...  \
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
25794  76561198306599751  http://steamcommunity.com/profiles/76561198306...   
25795           Ghoustik              http://steamcommunity.com/id/Ghoustik   
25796  76561198310819422  http://steamcommunity.com/profiles/76561198310...   
25797  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
25798        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  
0      [{'funny': '', 'posted': 'Posted November 5, 2...  
1      [{'funny': '', 'posted': 'Posted June 24, 2014...  
2      [{'funny': '', 'posted': 'Posted February 3.',...  
3      [{'funny': '', 'posted': 'Posted October 14, 2...  
4      [{'funny': '3 people found this review funny',...  
...                                                  ...  
25794  [{'funny': '', 'posted': 'Posted May 31.', 'la...  
25795  [{'funny': '', 'posted': 'Posted June 17.', 'l...  
25796  [{'funny': '1 person found this review funny',...  
25797  [{'funny': '', 'posted': 'Posted July 21.', 'l...  
25798  [{'funny': '1 person found this review funny',...  

[25799 rows x 3 columns]

In [5]:
rows = []

with gzip.open("users_items.json.gz", mode='rt', encoding="utf-8") as archivo:
    for linea in archivo:
        try:
            obj = ast.literal_eval(linea)
            rows.append(obj)
        except json.JSONDecodeError:
            print(f'Error en la linea {linea}')


user_items = pd.DataFrame(rows)

user_items

user_id  items_count           steam_id   
0            76561197970982479          277  76561197970982479  \
1                      js41637          888  76561198035864385   
2                    evcentric          137  76561198007712555   
3                   Riot-Punch          328  76561197963445855   
4                        doctr          541  76561198002099482   
...                        ...          ...                ...   
88305        76561198323066619           22  76561198323066619   
88306        76561198326700687          177  76561198326700687   
88307  XxLaughingJackClown77xX            0  76561198328759259   
88308        76561198329548331            7  76561198329548331   
88309        edward_tremethick            0  76561198331598578   

                                                user_url   
0      http://steamcommunity.com/profiles/76561197970...  \
1                   http://steamcommunity.com/id/js41637   
2                 http://steamcommunity.com/id/evcentric   
3                http://steamcommunity.com/id/Riot-Punch   
4                     http://steamcommunity.com/id/doctr   
...                                                  ...   
88305  http://steamcommunity.com/profiles/76561198323...   
88306  http://steamcommunity.com/profiles/76561198326...   
88307  http://steamcommunity.com/id/XxLaughingJackClo...   
88308  http://steamcommunity.com/profiles/76561198329...   
88309     http://steamcommunity.com/id/edward_tremethick   

                                                   items  
0      [{'item_id': '10', 'item_name': 'Counter-Strik...  
1      [{'item_id': '10', 'item_name': 'Counter-Strik...  
2      [{'item_id': '1200', 'item_name': 'Red Orchest...  
3      [{'item_id': '10', 'item_name': 'Counter-Strik...  
4      [{'item_id': '300', 'item_name': 'Day of Defea...  
...                                                  ...  
88305  [{'item_id': '413850', 'item_name': 'CS:GO Pla...  
88306  [{'item_id': '11020', 'item_name': 'TrackMania...  
88307                                                 []  
88308  [{'item_id': '304930', 'item_name': 'Unturned'...  
88309                                                 []  

[88310 rows x 5 columns]

#### ETL primer dataset (steam_games)

In [6]:
dfgames = steam_games.dropna() #borramos todas las filas que tengan Nan (Se opta por esto para reducir el dataset para render, lo correcto seria borrar donde todas las filas tiene Nan)

dfgames.drop(columns=['url', 'reviews_url', 'specs', 'early_access', 'publisher'], inplace=True) #Eliminamos las columnas que no vamos a utilizar



C:\Users\Javo\AppData\Local\Temp\ipykernel_420\46518938.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfgames.drop(columns=['url', 'reviews_url', 'specs', 'early_access', 'publisher'], inplace=True) #Eliminamos las columnas que no vamos a utilizar


In [7]:
dfgames.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22530 entries, 88310 to 120443
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   genres        22530 non-null  object 
 1   app_name      22530 non-null  object 
 2   title         22530 non-null  object 
 3   release_date  22530 non-null  object 
 4   tags          22530 non-null  object 
 5   price         22530 non-null  object 
 6   id            22530 non-null  float64
 7   developer     22530 non-null  object 
dtypes: float64(1), object(7)
memory usage: 1.5+ MB


In [8]:
dfgames

genres   
88310       [Action, Casual, Indie, Simulation, Strategy]  \
88311                [Free to Play, Indie, RPG, Strategy]   
88312   [Casual, Free to Play, Indie, Simulation, Sports]   
88313                         [Action, Adventure, Casual]   
88315                     [Action, Adventure, Simulation]   
...                                                   ...   
120439                 [Action, Adventure, Casual, Indie]   
120440              [Casual, Indie, Simulation, Strategy]   
120441                          [Casual, Indie, Strategy]   
120442                        [Indie, Racing, Simulation]   
120443                                    [Casual, Indie]   

                        app_name                     title release_date   
88310        Lost Summoner Kitty       Lost Summoner Kitty   2018-01-04  \
88311                  Ironbound                 Ironbound   2018-01-04   
88312    Real Pool 3D - Poolians   Real Pool 3D - Poolians   2017-07-24   
88313                    弹炸人2222                   弹炸人2222   2017-12-07   
88315      Battle Royale Trainer     Battle Royale Trainer   2018-01-04   
...                          ...                       ...          ...   
120439              Kebab it Up!              Kebab it Up!   2018-01-04   
120440            Colony On Mars            Colony On Mars   2018-01-04   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   2018-01-04   
120442             Russian Roads             Russian Roads   2018-01-04   
120443       EXIT 2 - Directions       EXIT 2 - Directions   2017-09-02   

                                                     tags         price   
88310       [Strategy, Action, Indie, Casual, Simulation]          4.99  \
88311   [Free to Play, Strategy, Indie, RPG, Card Game...  Free To Play   
88312   [Free to Play, Simulation, Sports, Casual, Ind...  Free to Play   
88313                         [Action, Adventure, Casual]          0.99   
88315   [Action, Adventure, Simulation, FPS, Shooter, ...          3.99   
...                                                   ...           ...   
120439        [Action, Indie, Casual, Violent, Adventure]          1.99   
120440              [Strategy, Indie, Casual, Simulation]          1.99   
120441                          [Strategy, Indie, Casual]          4.99   
120442                        [Indie, Simulation, Racing]          1.99   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...          4.99   

              id                 developer  
88310   761140.0                 Kotoshiro  
88311   643980.0          Secret Level SRL  
88312   670290.0              Poolians.com  
88313   767400.0                      彼岸领域  
88315   772540.0       Trickjump Games Ltd  
...          ...                       ...  
120439  745400.0           Bidoniera Games  
120440  773640.0        Nikita "Ghost_RUS"  
120441  733530.0                    Sacada  
120442  610660.0  Laush Dmitriy Sergeevich  
120443  658870.0             xropi,stev3ns  

[22530 rows x 8 columns]

In [9]:
#Juntamos app_name y title en otra columna, la insertamos y borramos las columnas de origen
val = []
for i, e in enumerate(dfgames['app_name']):
    if e == None:
        val.append(dfgames['title'].iloc[i])
    else:
        val.append(e)


dfgames.insert(3, 'titles', val)

dfgames.drop(columns=['app_name', 'title'], inplace=True)

C:\Users\Javo\AppData\Local\Temp\ipykernel_420\1365523421.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfgames.drop(columns=['app_name', 'title'], inplace=True)


In [10]:
#Repetimos procedimiento con genres y tags. Tags tiene más etiquetas y la busqueda puede ser más específica
#si se quiere buscar algo fuera de los generos tradicionales
main = []

for i in dfgames['genres']:
    main.append(i[0])
        
val = []
for i, e in enumerate(dfgames['tags']):
    if e == None:
        val.append(dfgames['genres'].iloc[i])
    else:
        val.append(e)
    
    

dfgames.insert(1, 'genre', val)


dfgames.drop(columns=['genres', 'tags'], inplace=True)

C:\Users\Javo\AppData\Local\Temp\ipykernel_420\2580807934.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfgames.drop(columns=['genres', 'tags'], inplace=True)


In [11]:
#La lista de etiquetas la pasamos a string
dfgames['genres'] = dfgames['genre'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)

C:\Users\Javo\AppData\Local\Temp\ipykernel_420\2373711721.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfgames['genres'] = dfgames['genre'].apply(lambda x: ', '.join(x) if isinstance(x, list) else x)


In [12]:
#borramos la columna utilizada para la funcion anterior
dfgames.drop(columns='genre', inplace=True)

C:\Users\Javo\AppData\Local\Temp\ipykernel_420\3733617105.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfgames.drop(columns='genre', inplace=True)


In [13]:
#Observamos valores en str, por lo tanto vamos a reemplazar esos valores por 0 por falta de informacion
#para que no afecte la suma en las consultas
dfgames['price'].unique()

array([4.99, 'Free To Play', 'Free to Play', 0.99, 3.99, 18.99, 29.99,
       10.99, 2.99, 1.5899999999999999, 14.99, 1.99, 59.99, 9.99, 6.99,
       7.99, 39.99, 'Free', 19.99, 7.49, 8.99, 12.99, 5.99, 2.49, 15.99,
       1.25, 24.99, 17.99, 61.99, 3.49, 11.99, 13.99, 'Free Demo',
       'Play for Free!', 34.99, 1.49, 32.99, 99.99, 14.95, 69.99, 16.99,
       79.99, 49.99, 13.98, 149.99, 771.71, 'Install Now', 21.99, 89.99,
       'Play WARMACHINE: Tactics Demo', 0.98, 139.92, 4.29, 'Free Mod',
       54.99, 64.99, 74.99, 0.89, 0.5, 'Play Now', 299.99, 1.29, 119.99,
       44.99, 3.0, 15.0, 1.3900000000000001, 'Free HITMAN™ Holiday Pack',
       2.0, 4.0, 1.9500000000000002, 1.5, 6.66, 26.99, 399.99, 31.99,
       20.0, 40.0, 5.0, 3.33, 38.85, 71.7, 995.0, 5.49, 27.49, 3.39, 6.0,
       19.95, 20.99, 499.99, 27.99, 199.99, 4.68, 131.4, 44.98, 202.76,
       2.3, 0.9500000000000001, 36.99, 172.24, 249.99, 2.9699999999999998,
       10.96, 2.66, 6.48, 10.0, 1.0, 11.15, 'Play the Demo', 

In [14]:
dfgames['price'].fillna(0, inplace=True) #llenamos valores vacios con 0
dfgames['price'] = dfgames['price'].apply(lambda x: 0 if type(x) == str else float(x))


C:\Users\Javo\AppData\Local\Temp\ipykernel_420\1976791066.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfgames['price'].fillna(0, inplace=True) #llenamos valores vacios con 0
C:\Users\Javo\AppData\Local\Temp\ipykernel_420\1976791066.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfgames['price'] = dfgames['price'].apply(lambda x: 0 if type(x) == str else float(x))


In [15]:
#renombro la columna para luego poder hacer un merge
dfgames.rename(columns={'id':'item_id'}, inplace=True)

C:\Users\Javo\AppData\Local\Temp\ipykernel_420\1398374538.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfgames.rename(columns={'id':'item_id'}, inplace=True)


In [16]:
#Cambiamos el tipo de dato de la columna
dfgames['item_id'] = dfgames['item_id'].astype('int')
dfgames.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22530 entries, 88310 to 120443
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   titles        22530 non-null  object 
 1   release_date  22530 non-null  object 
 2   price         22530 non-null  float64
 3   item_id       22530 non-null  int32  
 4   developer     22530 non-null  object 
 5   genres        22530 non-null  object 
dtypes: float64(1), int32(1), object(4)
memory usage: 1.1+ MB


C:\Users\Javo\AppData\Local\Temp\ipykernel_420\456471097.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfgames['item_id'] = dfgames['item_id'].astype('int')


In [17]:
#Pasamos todas las fechas a año
import re

patron = r'(\d{4})'  # Este patrón busca 4 dígitos consecutivos en una cadena cruda (r: raw).

def año(cadena:str) -> str|None:
    """
    La función año toma la cadena de texto y devuelve un año en formato entero
    o None si la cadena está vacía.
    """
    match = re.search(patron, str(cadena))
    if match:
        return match.group(1)
    else:
        return None

dfgames['release_date'] = dfgames['release_date'].apply(año)

C:\Users\Javo\AppData\Local\Temp\ipykernel_420\2525101533.py:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfgames['release_date'] = dfgames['release_date'].apply(año)


In [18]:
#Pasamos a lowercase y le quitamos los espacios sobrantes a las columnas con str
dfgames['titles'] = dfgames['titles'].apply(lambda x: x.lower().strip())
dfgames['developer'] = dfgames['developer'].apply(lambda x: x.lower().strip())
dfgames['genres'] = dfgames['genres'].apply(lambda x: x.lower().strip())
dfgames

C:\Users\Javo\AppData\Local\Temp\ipykernel_420\3834056437.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfgames['titles'] = dfgames['titles'].apply(lambda x: x.lower().strip())
C:\Users\Javo\AppData\Local\Temp\ipykernel_420\3834056437.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfgames['developer'] = dfgames['developer'].apply(lambda x: x.lower().strip())
C:\Users\Javo\AppData\Local\Temp\ipykernel_420\3834056437.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice fr

titles release_date  price  item_id   
88310        lost summoner kitty         2018   4.99   761140  \
88311                  ironbound         2018   0.00   643980   
88312    real pool 3d - poolians         2017   0.00   670290   
88313                    弹炸人2222         2017   0.99   767400   
88315      battle royale trainer         2018   3.99   772540   
...                          ...          ...    ...      ...   
120439              kebab it up!         2018   1.99   745400   
120440            colony on mars         2018   1.99   773640   
120441  logistical: south africa         2018   4.99   733530   
120442             russian roads         2018   1.99   610660   
120443       exit 2 - directions         2017   4.99   658870   

                       developer   
88310                  kotoshiro  \
88311           secret level srl   
88312               poolians.com   
88313                       彼岸领域   
88315        trickjump games ltd   
...                          ...   
120439           bidoniera games   
120440        nikita "ghost_rus"   
120441                    sacada   
120442  laush dmitriy sergeevich   
120443             xropi,stev3ns   

                                                   genres  
88310         strategy, action, indie, casual, simulation  
88311   free to play, strategy, indie, rpg, card game,...  
88312   free to play, simulation, sports, casual, indi...  
88313                           action, adventure, casual  
88315   action, adventure, simulation, fps, shooter, t...  
...                                                   ...  
120439          action, indie, casual, violent, adventure  
120440                strategy, indie, casual, simulation  
120441                            strategy, indie, casual  
120442                          indie, simulation, racing  
120443  indie, casual, puzzle, singleplayer, atmospher...  

[22530 rows x 6 columns]

#### ETL segundo dataset (user_reviews)

In [19]:
dfreviews = user_reviews.dropna(how='all') #borramos todas las filas donde todas las columnas tengan Nan

dfreviews.drop(columns=['user_url'], inplace=True) #Eliminamos la columna de url ya que no las vamos a utilizar

In [20]:
#La copiamos porque más adelante la utilizamos
dfreviews2 = dfreviews.copy()

In [21]:
#Hacemos una lista con las review de los distintos juegos
lista = []
for elem in dfreviews['reviews']:
    for e in elem:
        l = [e['item_id'], e['review']]
        lista.append(l)


In [22]:
#importamos la herramienta para hacer el analisis de sentimiento
import nltk
from nltk.sentiment.vader import SentimentIntensityAnalyzer

#Descarga de recursos necesarios para VADER
nltk.download('vader_lexicon')

#Creación del objeto SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\Javo\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [23]:
dfsent = pd.DataFrame(lista, columns=['item_id', 'review']) #creamos un dataframe nuevo donde vamos a hacer el analisis de sentimiento

In [24]:
#le añadimos el puntaje de sentimiento a cada review
dfsent['sentiment'] = dfsent['review'].apply(lambda x: analyzer.polarity_scores(x)['compound'])

In [25]:
#Agregamos columnas con 1 y 0 por cada sentimiento para luego poder hacer el conteo
dfsent['sent_neg'] = dfsent['sentiment'].apply(lambda x: 1 if x < 0 else 0)
dfsent['sent_neu'] = dfsent['sentiment'].apply(lambda x: 1 if 0 <= x <= 0.3 else 0)
dfsent['sent_pos'] = dfsent['sentiment'].apply(lambda x: 1 if x > 0.3 else 0)

In [26]:
#Agrupamos por id y nos queda un conteo de la cantidad de reviews que tiene cada juego
dfsent.drop(columns=['review', 'sentiment'], inplace=True)
dfsent = dfsent.groupby('item_id').sum()

In [27]:
#Aca le agregamos la columna de sentimiento en funcion de la cantidad de review de cada tipo
def colsent(df):
    if df['sent_pos'] > df["sent_neu"] and df['sent_pos'] > df["sent_neg"]:
        return 2
    elif df['sent_neu'] >= df["sent_pos"] and df['sent_neu'] >= df["sent_neg"]:
        return 1
    else:
        return 0
    
dfsent['sentiment'] = dfsent.apply(colsent, axis=1)

In [28]:
#Reseteamos el indice y pasamos item_id a int
dfsent = dfsent.reset_index()
dfsent['item_id'] = dfsent['item_id'].astype('int')

In [29]:
dfsent

item_id  sent_neg  sent_neu  sent_pos  sentiment
0          10         2        25        30          2
1       10090         6         7        39          2
2       10130         0         0         2          2
3       10140         1         0         0          0
4       10150         1         1         7          2
...       ...       ...       ...       ...        ...
3677    99400         1         0         0          0
3678    99700         0         1         2          2
3679    99810         2         1         5          2
3680    99900        10        55        89          2
3681    99910         1         0         3          2

[3682 rows x 5 columns]

In [30]:
#Hacemos una lista con los recommend de cada usuario
lista = []
for i in range(0, len(dfreviews)):
    l = []
    l.append(dfreviews['user_id'][i])
    for elem in dfreviews['reviews'][i]:
        l.append(elem['recommend'])
    lista.append(l)


In [31]:
# Extraemos la cantidad de reviews positivas y negativas por usuario
pos = []
neg = []

for i in range(len(lista)):
    x = 0
    z = 0
    for elem in lista[i]:
        if elem == True:
            x = x + 1
        if elem == False:
            z = z + 1
    pos.append(x)
    neg.append(z)


In [32]:
dfreviews['recommend_pos'] = pos
dfreviews['recommend_neg'] = neg 

In [33]:
#Aca le agregamos la columna con el porcentaje de recomendacion de cada usuario
def percent(df):
    if df['recommend_pos'] == 0 and df['recommend_neg'] == 0:
        return 0
    else:
        return round(df['recommend_pos'] / (df['recommend_pos'] + df['recommend_neg']) * 100, 2)
    
    
dfreviews["% recommend"] = dfreviews.apply(percent, axis=1)

In [34]:
#Una vez extraido todo lo necesario, borramos la columna de reviews
dfreviews.drop(columns='reviews', inplace=True)

In [35]:
dfreviews

user_id  recommend_pos  recommend_neg  % recommend
0      76561197970982479              3              0        100.0
1                js41637              3              0        100.0
2              evcentric              6              0        100.0
3                  doctr              6              0        100.0
4              maplemage              4              0        100.0
...                  ...            ...            ...          ...
25794  76561198306599751              1              0        100.0
25795           Ghoustik              1              0        100.0
25796  76561198310819422              1              0        100.0
25797  76561198312638244              4              0        100.0
25798        LydiaMorley              3              0        100.0

[25799 rows x 4 columns]

In [36]:
lista = []
for i in dfreviews2['reviews']:
    for e in i:
        if e['recommend'] == True:
            l = [e['item_id'], 1]
        else:
            l = [e['item_id'], 0]
            
        lista.append(l)

In [37]:
#calculamos cantidad de recomendaciones de cada juego
df = pd.DataFrame(lista, columns=['item_id', 'recommend_True']).groupby('item_id').sum()
df = df.reset_index()
df['item_id'] = dfsent['item_id'].astype('int')


In [38]:
dfsent = pd.merge(df, dfsent, on='item_id')

In [39]:
dfsent 

item_id  recommend_True  sent_neg  sent_neu  sent_pos  sentiment
0          10              56         2        25        30          2
1       10090              51         6         7        39          2
2       10130               2         0         0         2          2
3       10140               1         1         0         0          0
4       10150               9         1         1         7          2
...       ...             ...       ...       ...       ...        ...
3677    99400               0         1         0         0          0
3678    99700               3         0         1         2          2
3679    99810               5         2         1         5          2
3680    99900             137        10        55        89          2
3681    99910               3         1         0         3          2

[3682 rows x 6 columns]

### ETL tercer dataset (user_items)

In [40]:
dfitems = user_items.dropna() #Eliminamos nulos
dfitems.drop(columns=['user_url', 'steam_id'], inplace=True) #Eliminamos columnas que no vamos a usar

In [41]:
dfitems

user_id  items_count   
0            76561197970982479          277  \
1                      js41637          888   
2                    evcentric          137   
3                   Riot-Punch          328   
4                        doctr          541   
...                        ...          ...   
88305        76561198323066619           22   
88306        76561198326700687          177   
88307  XxLaughingJackClown77xX            0   
88308        76561198329548331            7   
88309        edward_tremethick            0   

                                                   items  
0      [{'item_id': '10', 'item_name': 'Counter-Strik...  
1      [{'item_id': '10', 'item_name': 'Counter-Strik...  
2      [{'item_id': '1200', 'item_name': 'Red Orchest...  
3      [{'item_id': '10', 'item_name': 'Counter-Strik...  
4      [{'item_id': '300', 'item_name': 'Day of Defea...  
...                                                  ...  
88305  [{'item_id': '413850', 'item_name': 'CS:GO Pla...  
88306  [{'item_id': '11020', 'item_name': 'TrackMania...  
88307                                                 []  
88308  [{'item_id': '304930', 'item_name': 'Unturned'...  
88309                                                 []  

[88310 rows x 3 columns]

In [42]:
dfitems['items'][0]

[{'item_id': '10',
  'item_name': 'Counter-Strike',
  'playtime_forever': 6,
  'playtime_2weeks': 0},
 {'item_id': '20',
  'item_name': 'Team Fortress Classic',
  'playtime_forever': 0,
  'playtime_2weeks': 0},
 {'item_id': '30',
  'item_name': 'Day of Defeat',
  'playtime_forever': 7,
  'playtime_2weeks': 0},
 {'item_id': '40',
  'item_name': 'Deathmatch Classic',
  'playtime_forever': 0,
  'playtime_2weeks': 0},
 {'item_id': '50',
  'item_name': 'Half-Life: Opposing Force',
  'playtime_forever': 0,
  'playtime_2weeks': 0},
 {'item_id': '60',
  'item_name': 'Ricochet',
  'playtime_forever': 0,
  'playtime_2weeks': 0},
 {'item_id': '70',
  'item_name': 'Half-Life',
  'playtime_forever': 0,
  'playtime_2weeks': 0},
 {'item_id': '130',
  'item_name': 'Half-Life: Blue Shift',
  'playtime_forever': 0,
  'playtime_2weeks': 0},
 {'item_id': '300',
  'item_name': 'Day of Defeat: Source',
  'playtime_forever': 4733,
  'playtime_2weeks': 0},
 {'item_id': '240',
  'item_name': 'Counter-Strike: S

In [43]:
#Hacemos un dataframe con toda la informacion 
lista = []

for i in range(len(dfitems["items"])):
    user_id = dfitems["user_id"][i]
    for j in dfitems["items"][i]:
        diccionario_items = j.copy()
        diccionario_items["user_id"] = user_id
        lista.append(diccionario_items)



In [44]:
itemsd = pd.DataFrame(lista)

In [45]:
itemsd

item_id                      item_name  playtime_forever   
0            10                 Counter-Strike                 6  \
1            20          Team Fortress Classic                 0   
2            30                  Day of Defeat                 7   
3            40             Deathmatch Classic                 0   
4            50      Half-Life: Opposing Force                 0   
...         ...                            ...               ...   
5153204  346330                   BrainBread 2                 0   
5153205  373330                    All Is Dust                 0   
5153206  388490  One Way To Die: Steam Edition                 3   
5153207  521570          You Have 10 Seconds 2                 4   
5153208  519140                     Minds Eyes                 3   

         playtime_2weeks            user_id  
0                      0  76561197970982479  
1                      0  76561197970982479  
2                      0  76561197970982479  
3                      0  76561197970982479  
4                      0  76561197970982479  
...                  ...                ...  
5153204                0  76561198329548331  
5153205                0  76561198329548331  
5153206                3  76561198329548331  
5153207                4  76561198329548331  
5153208                3  76561198329548331  

[5153209 rows x 5 columns]

In [46]:
#Eliminamos la columna que no vamos a utilizar
itemsd.drop(columns='playtime_2weeks', inplace=True)

In [47]:
#Transformamos la columna a tipo entero
itemsd['item_id'] = itemsd['item_id'].astype('int')

In [48]:
itemsd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5153209 entries, 0 to 5153208
Data columns (total 4 columns):
 #   Column            Dtype 
---  ------            ----- 
 0   item_id           int32 
 1   item_name         object
 2   playtime_forever  int64 
 3   user_id           object
dtypes: int32(1), int64(1), object(2)
memory usage: 137.6+ MB


In [49]:
itemsd

item_id                      item_name  playtime_forever   
0             10                 Counter-Strike                 6  \
1             20          Team Fortress Classic                 0   
2             30                  Day of Defeat                 7   
3             40             Deathmatch Classic                 0   
4             50      Half-Life: Opposing Force                 0   
...          ...                            ...               ...   
5153204   346330                   BrainBread 2                 0   
5153205   373330                    All Is Dust                 0   
5153206   388490  One Way To Die: Steam Edition                 3   
5153207   521570          You Have 10 Seconds 2                 4   
5153208   519140                     Minds Eyes                 3   

                   user_id  
0        76561197970982479  
1        76561197970982479  
2        76561197970982479  
3        76561197970982479  
4        76561197970982479  
...                    ...  
5153204  76561198329548331  
5153205  76561198329548331  
5153206  76561198329548331  
5153207  76561198329548331  
5153208  76561198329548331  

[5153209 rows x 4 columns]

In [50]:
#Unimos dos df para que para que nos queden los datos de los juegos en el mismo df
itemsd = pd.merge(itemsd,dfgames, on='item_id', how='inner')

In [51]:
itemsd.drop(columns=['titles', 'price', 'developer'], inplace=True)

In [52]:
#Borramos todos los juegos que no tienen horas jugadas porque no van a sumar a la hora de hacer la consulta
itemsd.drop(itemsd[(itemsd['playtime_forever'] == 0)].index, inplace=True)

In [53]:
itemsd['item_name'] = itemsd['item_name'].apply(lambda x: x.lower().strip())
itemsd

item_id                                     item_name   
0             10                                counter-strike  \
3             10                                counter-strike   
4             10                                counter-strike   
7             10                                counter-strike   
8             10                                counter-strike   
...          ...                                           ...   
4039494   433400                                           kim   
4039495   456780                     ducati - 90th anniversary   
4039496   444770                                  mimpi dreams   
4039497   375450  nobunaga's ambition: souzou sengokurisshiden   
4039498   354280                                    chaostower   

         playtime_forever            user_id release_date   
0                       6  76561197970982479         2000  \
3                      93              doctr         2000   
4                     108     corrupted_soul         2000   
7                     328         WeiEDKrSat         2000   
8                    6275       death-hunter         2000   
...                   ...                ...          ...   
4039494                13      thekushcorner         2016   
4039495                 6       peter1023199         2016   
4039496                46  76561198071808318         2016   
4039497                51  76561198081168593         2016   
4039498               164  76561198107283457         2016   

                                                    genres  
0        action, fps, multiplayer, shooter, classic, te...  
3        action, fps, multiplayer, shooter, classic, te...  
4        action, fps, multiplayer, shooter, classic, te...  
7        action, fps, multiplayer, shooter, classic, te...  
8        action, fps, multiplayer, shooter, classic, te...  
...                                                    ...  
4039494  rpg, indie, adventure, procedural generation, ...  
4039495                         simulation, racing, sports  
4039496  indie, adventure, cute, puzzle-platformer, pla...  
4039497                               strategy, historical  
4039498                  indie, casual, action, simulation  

[2708529 rows x 6 columns]

# Preparacion de datasets para los endpoints

Una vez que tenemos los datos listos, vamos a preparar un dataset optimizado para cada endpoint para economizar los recursos de Render

1° endpoint

def developer( desarrollador : str ): Cantidad de items y porcentaje de contenido Free por año según empresa desarrolladora.

In [54]:
end1 = dfgames.copy()
end1.drop(columns=['genres','titles', 'item_id'], inplace=True)

In [55]:
#creamos una columna para identidicar rapidamente los FreetoPlay
end1['free'] = end1['price'].apply(lambda x: 1 if x == 0 else 0)

In [56]:
#Borramos price y exportamos la data para usar
end1.drop(columns='price', inplace=True)
#end1.to_parquet('endpoint1', compression='snappy')

In [57]:
end1

release_date                 developer  free
88310          2018                 kotoshiro     0
88311          2018          secret level srl     1
88312          2017              poolians.com     1
88313          2017                      彼岸领域     0
88315          2018       trickjump games ltd     0
...             ...                       ...   ...
120439         2018           bidoniera games     0
120440         2018        nikita "ghost_rus"     0
120441         2018                    sacada     0
120442         2018  laush dmitriy sergeevich     0
120443         2017             xropi,stev3ns     0

[22530 rows x 3 columns]

Endpoint 2

def userdata( User_id : str ): Debe devolver cantidad de dinero gastado por el usuario, el porcentaje de recomendación en base a reviews.recommend y cantidad de items.

In [58]:
df =itemsd.dropna() #Eliminamos nulos
df['item_id'] = df['item_id'].astype('int')

In [59]:
#Calculamos cantidad de recomendaciones
df2 = dfreviews.drop(columns="% recommend")
df2 = df2.groupby(['user_id']).sum(['recommend_pos', 'recommend_neg']).reset_index()
df2['rec_games'] = df2[['recommend_pos', 'recommend_neg']].sum(axis=1)

In [60]:
#Hacemos merge para que las columnas que nos interesan de cada dataset queden en el mismo df y no pierdan correlación
end2 = pd.merge(df, dfgames, on='item_id', how='outer')
end2 = pd.merge(end2, df2, on='user_id', how='outer')
end2 = pd.merge(end2, dfitems, on='user_id', how='outer')

In [61]:
end2.drop(columns=['item_id', 'item_name', 'titles', 'release_date_x', 'developer', 'genres_x', 'recommend_pos',
                 'recommend_neg', 'items'], inplace=True)

In [62]:
end2.rename(columns={'release_date_y': 'release_date', 'genres_y' : 'genres'}, inplace=True)
end2

playtime_forever                  user_id release_date  price   
0                     6.0        76561197970982479         2000   9.99  \
1                     7.0        76561197970982479         2003   4.99   
2                  4733.0        76561197970982479         2010   9.99   
3                  1853.0        76561197970982479         2004  19.99   
4                   333.0        76561197970982479         2005   9.99   
...                   ...                      ...          ...    ...   
2808294               NaN        76561198316970597          NaN    NaN   
2808295               NaN        76561198318100691          NaN    NaN   
2808296               NaN        76561198326700687          NaN    NaN   
2808297               NaN  XxLaughingJackClown77xX          NaN    NaN   
2808298               NaN        edward_tremethick          NaN    NaN   

                                                    genres  rec_games   
0        action, fps, multiplayer, shooter, classic, te...        3.0  \
1        fps, world war ii, multiplayer, action, shoote...        3.0   
2        fps, world war ii, multiplayer, action, shoote...        3.0   
3        action, fps, multiplayer, shooter, team-based,...        3.0   
4        platformer, comedy, adventure, cult classic, a...        3.0   
...                                                    ...        ...   
2808294                                                NaN        NaN   
2808295                                                NaN        NaN   
2808296                                                NaN        NaN   
2808297                                                NaN        NaN   
2808298                                                NaN        NaN   

         items_count  
0              277.0  
1              277.0  
2              277.0  
3              277.0  
4              277.0  
...              ...  
2808294          0.0  
2808295          0.0  
2808296        177.0  
2808297          0.0  
2808298          0.0  

[2808299 rows x 7 columns]

In [63]:
end2.dropna(inplace=True ) #eliminamos los datos con muchos faltantes

In [64]:
end2['user_id'] = end2['user_id'].apply(lambda x: x.lower().strip())

In [65]:
end2

playtime_forever            user_id release_date  price   
0                     6.0  76561197970982479         2000   9.99  \
1                     7.0  76561197970982479         2003   4.99   
2                  4733.0  76561197970982479         2010   9.99   
3                  1853.0  76561197970982479         2004  19.99   
4                   333.0  76561197970982479         2005   9.99   
...                   ...                ...          ...    ...   
2773369               9.0    mrspecialperson         2013   0.00   
2773380               4.0  76561198083101258         2014   0.00   
2773381            1442.0  76561198083101258         2012   0.00   
2773476               6.0  76561198106773483         2013   9.99   
2773489             211.0        lldracuwulf         2014   4.99   

                                                    genres  rec_games   
0        action, fps, multiplayer, shooter, classic, te...        3.0  \
1        fps, world war ii, multiplayer, action, shoote...        3.0   
2        fps, world war ii, multiplayer, action, shoote...        3.0   
3        action, fps, multiplayer, shooter, team-based,...        3.0   
4        platformer, comedy, adventure, cult classic, a...        3.0   
...                                                    ...        ...   
2773369  pinball, simulation, free to play, arcade, cla...        1.0   
2773380  free to play, mmorpg, action, massively multip...        1.0   
2773381  anime, free to play, massively multiplayer, rp...        1.0   
2773476     adventure, horror, indie, action, first-person        1.0   
2773489  indie, adventure, action, platformer, sandbox,...        1.0   

         items_count  
0              277.0  
1              277.0  
2              277.0  
3              277.0  
4              277.0  
...              ...  
2773369          4.0  
2773380          4.0  
2773381          4.0  
2773476          1.0  
2773489         14.0  

[1425692 rows x 7 columns]

In [66]:
#end2 = end2.sample(n=50000)
#end2.to_parquet('endpoint2', compression='snappy')

Endpoint 3

def UserForGenre( genero : str ): Debe devolver el usuario que acumula más horas jugadas para el género dado y una lista de la acumulación de horas jugadas por año de lanzamiento.

Ejemplo de retorno: {"Usuario con más horas jugadas para Género X" : us213ndjss09sdf, "Horas jugadas":[{Año: 2013, Horas: 203}, {Año: 2012, Horas: 100}, {Año: 2011, Horas: 23}]}

In [67]:
itemsd

item_id                                     item_name   
0             10                                counter-strike  \
3             10                                counter-strike   
4             10                                counter-strike   
7             10                                counter-strike   
8             10                                counter-strike   
...          ...                                           ...   
4039494   433400                                           kim   
4039495   456780                     ducati - 90th anniversary   
4039496   444770                                  mimpi dreams   
4039497   375450  nobunaga's ambition: souzou sengokurisshiden   
4039498   354280                                    chaostower   

         playtime_forever            user_id release_date   
0                       6  76561197970982479         2000  \
3                      93              doctr         2000   
4                     108     corrupted_soul         2000   
7                     328         WeiEDKrSat         2000   
8                    6275       death-hunter         2000   
...                   ...                ...          ...   
4039494                13      thekushcorner         2016   
4039495                 6       peter1023199         2016   
4039496                46  76561198071808318         2016   
4039497                51  76561198081168593         2016   
4039498               164  76561198107283457         2016   

                                                    genres  
0        action, fps, multiplayer, shooter, classic, te...  
3        action, fps, multiplayer, shooter, classic, te...  
4        action, fps, multiplayer, shooter, classic, te...  
7        action, fps, multiplayer, shooter, classic, te...  
8        action, fps, multiplayer, shooter, classic, te...  
...                                                    ...  
4039494  rpg, indie, adventure, procedural generation, ...  
4039495                         simulation, racing, sports  
4039496  indie, adventure, cute, puzzle-platformer, pla...  
4039497                               strategy, historical  
4039498                  indie, casual, action, simulation  

[2708529 rows x 6 columns]

In [111]:
end3 = itemsd.copy()
end3.drop(columns=['item_name'], inplace=True)
end3['user_id'] = end3['user_id'].apply(lambda x: x.strip().lower())
end3

item_id  playtime_forever            user_id release_date   
0             10                 6  76561197970982479         2000  \
3             10                93              doctr         2000   
4             10               108     corrupted_soul         2000   
7             10               328         weiedkrsat         2000   
8             10              6275       death-hunter         2000   
...          ...               ...                ...          ...   
4039494   433400                13      thekushcorner         2016   
4039495   456780                 6       peter1023199         2016   
4039496   444770                46  76561198071808318         2016   
4039497   375450                51  76561198081168593         2016   
4039498   354280               164  76561198107283457         2016   

                                                    genres  
0        action, fps, multiplayer, shooter, classic, te...  
3        action, fps, multiplayer, shooter, classic, te...  
4        action, fps, multiplayer, shooter, classic, te...  
7        action, fps, multiplayer, shooter, classic, te...  
8        action, fps, multiplayer, shooter, classic, te...  
...                                                    ...  
4039494  rpg, indie, adventure, procedural generation, ...  
4039495                         simulation, racing, sports  
4039496  indie, adventure, cute, puzzle-platformer, pla...  
4039497                               strategy, historical  
4039498                  indie, casual, action, simulation  

[2708529 rows x 5 columns]

In [112]:
#Reducimos el dataset para poder cargarlo a render
end3 = end3.sample(n= 50000)
end3

item_id  playtime_forever            user_id release_date   
61516        240               190  76561198073626237         2004  \
3252096   253710                16  76561198086130805         2014   
2454959   257750               203            chocota         2013   
2986558   200170                77       oldleakyboat         2012   
1655478   214770               209            jayce76         2013   
...          ...               ...                ...          ...   
365531     10180                 1  76561198050560606         2009   
34366         70               117    majorsilentderp         1998   
2255000   102500               424          sgtcwispy         2012   
1737763   212480               414         ratchet719         2013   
1689945   252490              4987     crownnovacrown         2013   

                                                    genres  
61516    action, fps, multiplayer, shooter, team-based,...  
3252096  free to play, hunting, multiplayer, open world...  
2454959  platformer, indie, co-op, action, multiplayer,...  
2986558  strategy, multiplayer, funny, turn-based strat...  
1655478  metroidvania, platformer, action, indie, local...  
...                                                    ...  
365531   action, fps, multiplayer, shooter, first-perso...  
34366    fps, classic, action, sci-fi, singleplayer, sh...  
2255000  rpg, fantasy, open world, singleplayer, action...  
1737763  racing, local multiplayer, multiplayer, great ...  
1689945  early access, survival, crafting, multiplayer,...  

[50000 rows x 5 columns]

In [114]:
#Exportamos la data para usar
#end3.to_parquet('endpoint3', compression='snappy')

Endpoint 4 y 5

def best_developer_year( año : int ): Devuelve el top 3 de desarrolladores con juegos MÁS recomendados por usuarios para el año dado. (reviews.recommend = True y comentarios positivos)

Ejemplo de retorno: [{"Puesto 1" : X}, {"Puesto 2" : Y},{"Puesto 3" : Z}]

_____________________________________________________________

def developer_reviews_analysis( desarrolladora : str ): Según el desarrollador, se devuelve un diccionario con el nombre del desarrollador como llave y una lista con la cantidad total de registros de reseñas de usuarios que se encuentren categorizados con un análisis de sentimiento como valor positivo o negativo.

Ejemplo de retorno: {'Valve' : [Negative = 182, Positive = 278]}

In [73]:
#Con dfsent y dfgames hacemos un merge para que nos quede un df con todos los desarrolladores con juegos calificados
end4 = pd.merge(dfsent, dfgames, on='item_id')


In [74]:
#Eliminamos las columnas que no vamos a utilizar
end4.drop(columns=['item_id','sent_neu','titles','price','genres', 'sentiment'], inplace=True)

In [75]:
end4['release_date'] = end4['release_date'].astype('int')

In [76]:
end4

recommend_True  sent_neg  sent_pos  release_date              developer
0                 56         2        30          2000                  valve
1                 51         6        39          2008               treyarch
2                  2         0         2          2007      saber interactive
3                  1         1         0          2006           wanako games
4                  9         1         7          2009  radical entertainment
...              ...       ...       ...           ...                    ...
2887               3         0         2          2011      avalanche studios
2888               2         0         0          2010          krome studios
2889               0         1         0          2011              dead mage
2890               3         0         2          2011          nicalis, inc.
2891               3         1         3          2011            three rings

[2892 rows x 5 columns]

In [99]:
#Exportamos la data para usar
#end4.to_parquet('endpoint45', compression='snappy')

Sistema de recomendacion de juegos

In [77]:
dfsr = pd.merge(dfgames, dfsent, on='item_id', how='outer') #merge de los juegos con sus puntajes
dfsr.drop(columns=['release_date', 'price', 'sent_neu','sentiment', 'recommend_True'], inplace=True) #borramos cols que no vamos a usar
dfsr.dropna(subset='titles', inplace=True) #Borramos filas sin nombres de juegos
dfsr['sent_neg'].fillna(0, inplace=True) #llenamos los na de 0 en las filas de puntajes
dfsr['sent_pos'].fillna(0, inplace=True)
dfsr['titles'] = dfsr['titles'].apply(lambda x: x.strip()) #Borramos espacios y saltos de linea
dfsr['titles'] = dfsr['titles'].apply(lambda x: x.lower()) #ponemos todo en minuscula
dfsr.insert(3, 'main_genre', main)
dfsr

titles  item_id                 developer   
0           lost summoner kitty   761140                 kotoshiro  \
1                     ironbound   643980          secret level srl   
2       real pool 3d - poolians   670290              poolians.com   
3                       弹炸人2222   767400                      彼岸领域   
4         battle royale trainer   772540       trickjump games ltd   
...                         ...      ...                       ...   
22525              kebab it up!   745400           bidoniera games   
22526            colony on mars   773640        nikita "ghost_rus"   
22527  logistical: south africa   733530                    sacada   
22528             russian roads   610660  laush dmitriy sergeevich   
22529       exit 2 - directions   658870             xropi,stev3ns   

         main_genre                                             genres   
0            Action        strategy, action, indie, casual, simulation  \
1      Free to Play  free to play, strategy, indie, rpg, card game,...   
2            Casual  free to play, simulation, sports, casual, indi...   
3            Action                          action, adventure, casual   
4            Action  action, adventure, simulation, fps, shooter, t...   
...             ...                                                ...   
22525        Action          action, indie, casual, violent, adventure   
22526        Casual                strategy, indie, casual, simulation   
22527        Casual                            strategy, indie, casual   
22528         Indie                          indie, simulation, racing   
22529        Casual  indie, casual, puzzle, singleplayer, atmospher...   

       sent_neg  sent_pos  
0           0.0       0.0  
1           0.0       0.0  
2           0.0       0.0  
3           0.0       0.0  
4           0.0       0.0  
...         ...       ...  
22525       0.0       0.0  
22526       0.0       0.0  
22527       0.0       0.0  
22528       0.0       0.0  
22529       0.0       0.0  

[22530 rows x 7 columns]

In [90]:
#dfsr.to_parquet('recom_juegos', compression='snappy')

#Sampleamos para achicar el consumo de ram
#sample = dfsr.sample(n=10000)
#sample.to_parquet('recom_juegos', compression='snappy')

Segundo modelo de recomendacion

In [97]:
dfsr2 = pd.merge(itemsd, dfsr, on='item_id', how='inner')
dfsr2.drop(columns=['release_date','genres_y', 'item_name'], inplace=True)
dfsr2['user_id'] = dfsr2['user_id'].apply(lambda x: x.strip().lower())
dfsr2.rename(columns={'genres_x':'genres'}, inplace=True)
dfsr2

item_id  playtime_forever            user_id   
0             10                 6  76561197970982479  \
1             10                93              doctr   
2             10               108     corrupted_soul   
3             10               328         weiedkrsat   
4             10              6275       death-hunter   
...          ...               ...                ...   
2708524   433400                13      thekushcorner   
2708525   456780                 6       peter1023199   
2708526   444770                46  76561198071808318   
2708527   375450                51  76561198081168593   
2708528   354280               164  76561198107283457   

                                                    genres   
0        action, fps, multiplayer, shooter, classic, te...  \
1        action, fps, multiplayer, shooter, classic, te...   
2        action, fps, multiplayer, shooter, classic, te...   
3        action, fps, multiplayer, shooter, classic, te...   
4        action, fps, multiplayer, shooter, classic, te...   
...                                                    ...   
2708524  rpg, indie, adventure, procedural generation, ...   
2708525                         simulation, racing, sports   
2708526  indie, adventure, cute, puzzle-platformer, pla...   
2708527                               strategy, historical   
2708528                  indie, casual, action, simulation   

                                                    titles   
0                                           counter-strike  \
1                                           counter-strike   
2                                           counter-strike   
3                                           counter-strike   
4                                           counter-strike   
...                                                    ...   
2708524                                                kim   
2708525                          ducati - 90th anniversary   
2708526                                       mimpi dreams   
2708527  nobunaga's ambition: sphere of influence - asc...   
2708528                                         chaostower   

                          developer main_genre  sent_neg  sent_pos  
0                             valve     Action       2.0      30.0  
1                             valve     Action       2.0      30.0  
2                             valve     Action       2.0      30.0  
3                             valve     Action       2.0      30.0  
4                             valve     Action       2.0      30.0  
...                             ...        ...       ...       ...  
2708524    the secret games company  Adventure       0.0       0.0  
2708525            milestone s.r.l.     Racing       0.0       0.0  
2708526               silicon jelly  Adventure       0.0       0.0  
2708527  koei tecmo games co., ltd.   Strategy       0.0       0.0  
2708528                cashcowgames     Action       0.0       0.0  

[2708529 rows x 9 columns]

In [107]:
#Sampleamos para achicar el consumo de ram
#dfsr2 = dfsr2.sample(n=20000)
#dfsr2.to_parquet('recom_juegos2')
dfsr2

item_id  playtime_forever            user_id   
1678147     4580              1664  76561198064834139  \
1672253   207140              1837  76561198025816026   
178220       550                82         the_squash   
2330381   275100               110     frozentreasure   
2255257   206500               420            rakurix   
...          ...               ...                ...   
2400869    70600                16            daskrow   
2591032   285160               302  76561198035804388   
1714354   221100              1184  76561198099489942   
806468    346010                43      gamerz_united   
1777469   274290               393        coolman1342   

                                                    genres   
1678147  strategy, warhammer 40k, rts, sci-fi, multipla...  \
1672253  multiplayer, racing, local multiplayer, compet...   
178220   zombies, co-op, fps, multiplayer, action, onli...   
2330381  survival, open world, adventure, simulation, c...   
2255257  early access, free to play, strategy, multipla...   
...                                                    ...   
2400869  strategy, multiplayer, funny, turn-based, come...   
2591032  lego, adventure, action, open world, local co-...   
1714354  early access, survival, zombies, open world, m...   
806468   early access, building, sandbox, physics, dest...   
1777469  indie, pixel graphics, point & click, sci-fi, ...   

                                                 titles   
1678147  warhammer® 40,000: dawn of war® - dark crusade  \
1672253                                    speedrunners   
178220                                    left 4 dead 2   
2330381                                       shelter 2   
2255257                                  airmech strike   
...                                                 ...   
2400869                           worms ultimate mayhem   
2591032                               lego® the hobbit™   
1714354                                            dayz   
806468                                          besiege   
1777469                           gods will be watching   

                                         developer main_genre  sent_neg   
1678147                        relic entertainment   Strategy       0.0  \
1672253                          doubledutch games     Action      11.0   
178220                                       valve     Action     106.0   
2330381                          might and delight  Adventure       1.0   
2255257                               carbon games     Action      10.0   
...                                            ...        ...       ...   
2400869                         team17 digital ltd   Strategy       1.0   
2591032  traveller's tales,feral interactive (mac)     Action       3.0   
1714354                        bohemia interactive     Action     208.0   
806468                          spiderling studios      Indie      20.0   
1777469                             deconstructeam  Adventure       2.0   

         sent_pos  
1678147       8.0  
1672253      51.0  
178220      414.0  
2330381       1.0  
2255257      41.0  
...           ...  
2400869       2.0  
2591032       2.0  
1714354     373.0  
806468       36.0  
1777469       1.0  

[20000 rows x 9 columns]